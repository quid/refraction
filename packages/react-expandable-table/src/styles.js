/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import styled from '@emotion/styled/macro';
import css from '@emotion/css/macro';
import { GroupedVirtuoso } from 'react-virtuoso';
import { ThemeProvider, themes, textStyles } from '@quid/theme';
import { Icon, Text } from '@quid/react-core';
import { Container } from '@quid/react-tooltip';
import isPropValid from '@emotion/is-prop-valid';

const FLEX_ALIGN_MAP = {
  left: 'flex-start',
  right: 'flex-end',
  center: 'center',
};
const alignToFlex = (align = 'left') => FLEX_ALIGN_MAP[align];

export const ColumnCell = styled.div`
  box-sizing: border-box;
  flex-shrink: 0;
  display: flex;
  align-items: center;
  width: ${props => props.width};
  justify-content: ${props => alignToFlex(props.align)};
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  padding: 0 ${props => props.theme.sizes.small};
  ${props =>
    textStyles(...[props.bold ? 'bold' : null].filter(Boolean))(props)};
  color: ${themes.dark.primary};
`;
ColumnCell.defaultProps = {
  theme: themes.dark,
};

export const Row = styled.div`
  display: flex;
  align-items: stretch;
  position: relative;
  height: 40px;
  cursor: pointer;
  padding: 0 ${props => props.theme.sizes.regular};
  &::before,
  &::after {
    content: '';
    position: absolute;
    background-color: ${({ theme }) => theme.colors.gray5};
    transition: transform 0.2s ease-in-out;
  }
  &::before {
    top: 0;
    left: 0;
    bottom: 0;
    width: 4px;
    transform: scaleX(${({ open }) => (open ? 1 : 0)});
  }
  &::after {
    left: 0;
    bottom: 0;
    right: 0;
    height: 1px;
    transform: scaleY(${({ open }) => (open ? 1 : 0)});
  }
  &:not(:focus-visible) {
    outline: 0;
  }
  &:focus-visible {
    box-shadow: inset 0 0 0 1px ${props => props.theme.selected};
    z-index: 1;
  }
`;
Row.defaultProps = {
  theme: themes.dark,
};

export const ExpandableContent = styled.div`
  box-shadow: inset 0 -2px 2px -2px ${props => props.theme.colors.gray5};
  background-color: ${props => props.theme.colors.gray7};
`;
ExpandableContent.defaultProps = {
  theme: themes.dark,
};

export const List = styled(GroupedVirtuoso, {
  shouldForwardProp: prop =>
    ['group', 'item', 'groupCounts'].includes(prop) || isPropValid(prop),
})`
  color: ${props => props.theme.primary};
  background-color: ${props => props.theme.colors.gray7};
  border: 2px solid ${props => props.theme.colors.gray6};
  border-radius: 2px;
  max-height: ${props => (props.maxHeight ? `${props.maxHeight}px` : 'auto')};
`;
List.defaultProps = {
  theme: themes.dark,
};

export const Header = styled.header`
  display: flex;
  align-items: stretch;
  background-color: ${props => props.theme.colors.gray7};
  border-bottom: 1px solid ${props => props.theme.selected};
  border-radius: 2px 2px 0 0;
  padding: 0 ${props => props.theme.sizes.regular};
  height: 40px;
  font-weight: bold;

  ${Container} {
    font-weight: regular;
  }
`;
Header.defaultProps = {
  theme: themes.dark,
};

export const AngleButton = styled.button`
  all: unset;
  display: flex;
  align-items: center;
  justify-content: center;
  transform: rotate(${props => (props.open ? 180 : 0)}deg);
  transition: transform 0.5s ease-out;
  ${textStyles('xlarge')};
  border-radius: 50%;
  width: 20px;
  height: 20px;
  &:focus-visible {
    box-shadow: 0 0 2px 2px ${props => props.theme.selected};
  }
`;
AngleButton.defaultProps = {
  theme: themes.dark,
};

export const Ellipsis = styled(props => (
  <Text title={props.children} {...props} />
))`
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;
Ellipsis.defaultProps = {
  theme: themes.dark,
};

export const HeaderTitle = styled(Ellipsis)`
  margin-top: -2px;
  cursor: pointer;
  user-select: none;
  color: ${props =>
    props.inactive ? props.theme.colors.gray5 : props.theme.primary};
  &:focus:not(:focus-visible) {
    outline: 0;
  }
`;
HeaderTitle.defaultProps = {
  theme: themes.dark,
  tabIndex: -1,
};

export const InfoIcon = styled(
  React.forwardRef((props, ref) => (
    <button type="button" {...props} ref={ref}>
      <Icon name="question_circle" />
    </button>
  ))
)`
  all: unset;
  cursor: pointer;
  color: ${props => props.theme.colors.gray5};
  color: ${props => props.theme.prmary};
  margin-left: ${props => props.theme.sizes.small};
  flex-shrink: 0;

  &:hover,
  &:focus-visible {
    color: ${props => props.theme.primary};
  }
`;
InfoIcon.defaultProps = {
  theme: themes.dark,
};

const SortAsc = styled(props => <Icon name="sort_asc" {...props} />)``;
const SortDesc = styled(props => <Icon name="sort_desc" {...props} />)``;
const NEXT_ARROW = props =>
  props.sort == null ? SortDesc : props.sort === 'desc' ? SortAsc : false;
const ACTIVE_ARROW = props =>
  props.sort === 'desc' ? SortDesc : props.sort === 'asc' ? SortAsc : false;

export const SortIcon = styled(
  React.forwardRef(({ sort, ...props }, ref) => (
    <button type="button" {...props} ref={ref}>
      <SortAsc />
      <SortDesc />
    </button>
  ))
)`
  all: unset;
  display: inline-block;
  position: relative;
  height: 14px;
  width: 14px;
  cursor: pointer;
  margin-left: ${props => props.theme.sizes.small};
  flex-shrink: 0;

  ${SortAsc}, ${SortDesc} {
    position: absolute;
    top: 0;
    left: 0;
    color: ${props => props.theme.colors.gray5};
  }

  ${props =>
    ACTIVE_ARROW(props) &&
    css`
      ${ACTIVE_ARROW(props)} {
        color: ${props.theme.selected};
      }
    `}

  &:hover,
  &:focus-visible {
    ${props =>
      NEXT_ARROW(props) &&
      css`
        ${NEXT_ARROW(props)} {
          color: ${props.theme.primary};
        }
      `}
  }
`;
SortIcon.defaultProps = {
  theme: themes.dark,
};

export const TooltipContainer = styled(
  React.forwardRef(({ children, ...props }, ref) => (
    <ThemeProvider theme="light">
      <Container {...props} ref={ref}>
        {children}
      </Container>
    </ThemeProvider>
  ))
)`
  font-weight: normal;
  color: ${themes.light.primary};
`;
