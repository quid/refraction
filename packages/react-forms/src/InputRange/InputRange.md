The `InputRange` component aims to provide an interface compatible with [`input[type=range]`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range).

```js
<InputRange step={5} name="uncontrolled-with-step" onChange={console.log} />
```

All the properties supported by the HTML element are supported by `InputRange`, but,
additionally, this component supports two values at the same time.

The `onChange` handler in this case will return an object containing the properties
`start` and `end`, rather than a single number.

Rather than expecting a `value` property, this component expects a `values` one, which is an array of 1 or 2 numbers.
If only one number is provided, the component will render a single handle, when two are provided, the component will switch
to the two handles layout.

```js
<InputRange
  defaultValues={[10, 40]}
  name="uncontrolled-double"
  onChange={console.log}
/>
```

As with the HTML version, you can provide a `discrete` boolean property to display
some little ticks across the bar to show the valid and selectionable values.

```js
<InputRange
  step={5}
  defaultValues={[0]}
  name="uncontrolled-discrete"
  discrete
/>
```

This component can be used either as a [fully controlled][controlled-components]
or [fully uncontrolled][uncontrolled-components] component:

```js
initialState = {
  values: [0],
};

<InputRange
  values={state.values}
  name="controlled-single"
  onChange={values => setState({ values })}
  max={100}
/>;
```

[controlled-components]: https://reactjs.org/docs/forms.html#controlled-components
[uncontrolled-components]: https://reactjs.org/docs/uncontrolled-components.html
