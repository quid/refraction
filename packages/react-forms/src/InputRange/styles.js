/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import styled from '@emotion/styled/macro';
import { withFallback as wf } from '@quid/theme';

export const BASE_HANDLE_SIZE = 6;
export const HANDLE_SIZE = BASE_HANDLE_SIZE;
export const TICK_HEIGHT = 6;
export const TICK_WIDTH = 2;

export const TRACK_BACKGROUND = wf((props: Object) =>
  props.theme.current === 'light'
    ? props.theme.colors.gray4
    : props.theme.colors.gray2
);

export const Container = styled.div`
  position: relative;
  height: ${HANDLE_SIZE * 2}px;
  isolation: isolate;
`;

export const Track = styled.div`
  position: relative;
  top: 50%;
  background-position: center;
  background-size: 100% 2px;
  background-repeat: no-repeat;
  height: 2px;
  border-radius: 2px;
  transform: translateY(-50%);

  &::before {
    content: '';
    position: absolute;
    height: ${props => (props.disabled ? HANDLE_SIZE / 1.5 : HANDLE_SIZE)}px;
    width: 100%;
    transform: translateY(-25%);
  }
`;

export const Segment = styled.div`
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  right: 0;
  height: 100%;
  width: 100%;
  color: ${wf(props =>
    props.on ? TRACK_BACKGROUND(props) : props.theme.selected
  )};
`;

export const Tick = styled.div`
  height: ${TICK_HEIGHT}px;
  width: ${TICK_WIDTH}px;
  position: relative;
  transform: translateX(${TICK_WIDTH / 2}px) translateY(${TICK_HEIGHT / 2}px);
  color: ${wf(props => props.theme.selected)};
  border-radius: ${TICK_WIDTH}px;
  background-color: ${TRACK_BACKGROUND};
  opacity: ${wf(props => (props.hidden ? 0 : 1))};

  &:first-of-type {
    transform: translateY(${TICK_HEIGHT / 2}px);
  }
`;

export const StepTicks = styled(({ step, max, min, values, ...props }) => {
  const hidden = values
    .map(value => (value - min) / step)
    .filter(value => value !== undefined);

  return (
    <div {...props}>
      {Array(Math.ceil((max - min) / step) + 2)
        .fill()
        .map((_, key) => (
          <Tick key={key} hidden={hidden.includes(key)} />
        ))}
    </div>
  );
})`
  pointer-events: none;
  z-index: 0;
  display: flex;
  position: absolute;
  left: 0;
  top: -${BASE_HANDLE_SIZE - 1}px;
  right: 0;
  bottom: 0;
  justify-content: space-between;
`;

export const Rect = styled(props => (
  <rect
    {...props}
    fill="currentColor"
    height={2}
    width={Math.max(props.width, 0)}
    x={Math.max(props.offset, 0)}
    rx={1}
    ry={1}
  />
))`
  color: ${wf(props =>
    props.disabled ? TRACK_BACKGROUND(props) : props.theme.selected
  )};
`;
