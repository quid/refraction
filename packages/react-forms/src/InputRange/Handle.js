/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import styled from '@emotion/styled/macro';
import css from '@emotion/css/macro';
import { keyframes } from '@emotion/core';
import { withFallback as wf } from '@quid/theme';
import Color from 'color';
import { TRACK_BACKGROUND, HANDLE_SIZE } from './styles';

const pulseBg = props =>
  Color(props.theme.colors.quidTeal)
    .alpha(0.26)
    .string();

const pulse = wf(
  props => keyframes`
    from {
       box-shadow: 0 0 0 0 ${pulseBg(props)};
    }
    to {
      box-shadow: 0 0 0 ${HANDLE_SIZE / 2}px ${pulseBg(props)};
    }
`
);

const handleTransform = props => css`
  transform: scale(${props.disabled ? 2 : 3});
  transition-delay: 0s;
`;

type Props = {
  offset: number,
  dragging?: boolean,
  value?: number,
  name?: string,
  disabled?: boolean,
  readOnly?: boolean,
  zero?: boolean,
};

const Handle = styled(
  ({ dragging, value, name, disabled, readOnly, zero, ...props }: Props) => (
    <div
      tabIndex={!disabled ? 0 : undefined}
      data-action="drag-handle"
      data-value={value}
      {...props}
    >
      <input
        type="hidden"
        value={value}
        name={name}
        disabled={disabled}
        readOnly={readOnly}
      />
    </div>
  )
)`
  ${wf(
    props =>
      !(props.disabled || props.readOnly) &&
      css`
        color: ${props.theme.selected};
      `
  )}
  top: 50%;
  left: 0;
  position: absolute;
  width: ${HANDLE_SIZE}px;
  height: ${HANDLE_SIZE}px;
  margin-top: ${HANDLE_SIZE / -2}px;
  cursor: default;
  z-index: 2;
  &::before {
    content: '';
    display: block;
    background-color: ${wf(props =>
      props.disabled ? TRACK_BACKGROUND(props) : 'currentColor'
    )};
    width: ${HANDLE_SIZE}px;
    height: ${HANDLE_SIZE}px;
    transform: scale(2);
    border-radius: 50%;
    transition: transform 0.2s ease-in-out, width 0.2s ease-in-out,
      height 0.2s ease-in-out;
  }

  &:active::before {
    ${handleTransform};
  }

  ${props =>
    props.dragging &&
    css`
      &::before {
        ${handleTransform(props)}
      }
    `}

  &:focus {
    outline: 0;
  }

  &:focus-visible::before {
    animation: ${pulse} 1s infinite alternate ease-in-out;
  }
`;

export default Handle;
