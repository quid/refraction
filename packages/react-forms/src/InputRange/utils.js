/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import { Rect } from './styles';

export const fromPercentToPixel = (percent: number, width: number): number =>
  (width / 100) * percent;

export const drawSegment = (
  width: number,
  offset: number,
  disabled: boolean
) => (props: Object) =>
  wrapWithSvg(
    <Rect width={width} offset={offset} disabled={disabled} />,
    props
  );

export const wrapWithSvg = (
  content: React$Element<any>,
  { className, on, ...props }: { className?: string, on: boolean }
) => (
  <svg
    className={className}
    xmlns="http://www.w3.org/2000/svg"
    width="100"
    height="100"
    {...props}
  >
    {content}
  </svg>
);
