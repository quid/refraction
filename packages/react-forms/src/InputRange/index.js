/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import { Range } from 'react-range';
import useControlledState from '@quid/react-use-controlled-state';
import styled from '@emotion/styled/macro';
import useResizeAware from 'react-resize-aware';
import Handle from './Handle';
import { HANDLE_SIZE, Segment, StepTicks } from './styles';
import { fromPercentToPixel, drawSegment } from './utils';

type Props = {
  onChange?: (Array<number>) => void,
  values?: Array<number>,
  defaultValues?: Array<number>,
  name: string,
  min?: number,
  max?: number,
  step?: number,
  readOnly?: boolean,
  disabled?: boolean,
  discrete?: boolean,
  className?: string,
};
export const Track = styled.div`
  position: relative;
  top: 50%;
  background-position: center;
  background-size: 100% 2px;
  background-repeat: no-repeat;
  height: 2px;
  border-radius: 2px;
  transform: translateY(-50%);

  &::before {
    content: '';
    position: absolute;
    height: ${props => (props.disabled ? HANDLE_SIZE / 1.5 : HANDLE_SIZE)}px;
    width: 100%;
    transform: translateY(-25%);
  }
`;

const InputRange = styled(
  ({
    onChange,
    values: controlledValues,
    defaultValues,
    name,
    min = 0,
    max = 100,
    step = 1,
    readOnly = false,
    disabled = false,
    discrete = false,
    className,
    ...props
  }: Props) => {
    const [values = [min], setValues] = useControlledState(
      defaultValues,
      controlledValues,
      onChange
    );
    const [resizeListener, sizes] = useResizeAware();

    const [start, end = max] = values;
    const width = sizes.width;

    const getSegments = React.useCallback(
      (isTwoHandled: boolean) => {
        const isGrayedOut = disabled || readOnly;
        const unit = 100 / (max - min);

        const padding = isGrayedOut ? HANDLE_SIZE : 0;

        const a: number = isTwoHandled ? start - min : 0;
        const b: number = isTwoHandled ? end - min : start - min;

        const seg1W = fromPercentToPixel(unit * a, width);
        const Seg1 = drawSegment(seg1W - padding, 0, true);
        const seg2W =
          fromPercentToPixel(unit * b, width) -
          fromPercentToPixel(unit * a, width) -
          padding * (isTwoHandled ? 2 : 1);
        const seg2O = seg1W + (isTwoHandled ? padding : 0);
        const Seg2 = drawSegment(seg2W, seg2O, false);
        const Seg3 = drawSegment(
          fromPercentToPixel(100 - unit * b, width) - padding,
          fromPercentToPixel(unit * b, width) + padding,
          true
        );

        return [
          isTwoHandled && <Segment as={Seg1} key="1" />,
          <Segment as={Seg2} key="2" on />,
          <Segment as={Seg3} key="3" />,
        ];
      },
      [disabled, min, max, readOnly, width, start, end]
    );

    return (
      <Range
        values={values}
        onChange={setValues}
        min={min}
        max={max}
        step={step}
        disabled={disabled || readOnly}
        renderTrack={({ props, children }) => (
          <Track {...props} disabled={disabled} className={className}>
            {children}
            {getSegments(values.length === 2)}
            {discrete && (
              <StepTicks step={step} min={min} max={max} values={values} />
            )}
            {resizeListener}
          </Track>
        )}
        renderThumb={({ index, props, value }) => {
          const { transform, zIndex, ...styles } = props.style;
          return (
            <div
              style={{ position: 'absolute', zIndex: zIndex + 1, transform }}
            >
              <Handle
                {...props}
                style={styles}
                value={value}
                disabled={disabled}
                readOnly={readOnly}
                name={`${name}${index}`}
              />
            </div>
          );
        }}
        {...props}
      />
    );
  }
)();

export default InputRange;
