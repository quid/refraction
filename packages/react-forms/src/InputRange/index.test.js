/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import { mount } from 'enzyme';
import InputRange from '.';
import { Track, Rect, TRACK_BACKGROUND, HANDLE_SIZE } from './styles';
import Handle from './Handle';

jest.mock('react-resize-aware', () => () => [
  null,
  { width: 1000, height: 10 },
]);
jest.mock('react-range', () => ({
  __esModule: true,
  Range: ({ renderTrack, renderThumb, values, ...props }) => (
    <x-range {...props}>
      {renderTrack({ props: { style: { zIndex: 0 } } })}
      {values.map((value, index) =>
        renderThumb({ props: { style: { zIndex: 0 }, value, index } })
      )}
    </x-range>
  ),
}));

it('renders an InputRange', () => {
  expect(mount(<InputRange name="x" />)).toMatchSnapshot();
});

it('renders a disabled InputRange', () => {
  expect(
    mount(<InputRange name="x" disabled />)
      .find('rect')
      .at(1)
      .prop('width')
  ).toBe(994);
});

it('renders a discrete InputRange', () => {
  expect(
    mount(<InputRange name="x" discrete step={10} />).find('Tick')
  ).toHaveLength(12);
});

it('renders a double handle InputRange', () => {
  expect(
    mount(<InputRange name="x" defaultValues={[20, 80]} />).find('Handle')
  ).toHaveLength(2);
});

it('renders a Track component', () => {
  expect(mount(<Track />)).toMatchSnapshot();
  expect(mount(<Track disabled />)).toMatchSnapshot();
});

it('renders a disabled Handle', () => {
  expect(mount(<Handle disabled />).prop('tabIndex')).toBe(undefined);
});

it('renders a dragging Handle', () => {
  expect(mount(<Handle dragging />)).toHaveStyleRule('transition-delay', `0s`, {
    target: '::before',
  });
});

it('returns the alt background on dark theme', () => {
  expect(
    TRACK_BACKGROUND({
      theme: ({ current: 'dark', colors: { gray2: 'gray2' } }: any),
    })
  ).toMatchInlineSnapshot(`"gray2"`);
});

it('renders Rect', () => {
  // $FlowIgnoreMe
  console.error = jest.fn();
  expect(
    mount(<Rect width={-1} />)
      .find('rect')
      .prop('width')
  ).toBe(0);
});
