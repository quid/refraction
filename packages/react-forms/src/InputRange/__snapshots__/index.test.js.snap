// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Track component 1`] = `
.emotion-0 {
  position: relative;
  top: 50%;
  background-position: center;
  background-size: 100% 2px;
  background-repeat: no-repeat;
  height: 2px;
  border-radius: 2px;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.emotion-0::before {
  content: '';
  position: absolute;
  height: 6px;
  width: 100%;
  -webkit-transform: translateY(-25%);
  -ms-transform: translateY(-25%);
  transform: translateY(-25%);
}

<Track>
  <div
    className="emotion-0 emotion-1"
  />
</Track>
`;

exports[`renders a Track component 2`] = `
.emotion-0 {
  position: relative;
  top: 50%;
  background-position: center;
  background-size: 100% 2px;
  background-repeat: no-repeat;
  height: 2px;
  border-radius: 2px;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.emotion-0::before {
  content: '';
  position: absolute;
  height: 4px;
  width: 100%;
  -webkit-transform: translateY(-25%);
  -ms-transform: translateY(-25%);
  transform: translateY(-25%);
}

<Track
  disabled={true}
>
  <div
    className="emotion-0 emotion-1"
    disabled={true}
  />
</Track>
`;

exports[`renders an InputRange 1`] = `
@keyframes animation-0 {
  from {
    box-shadow: 0 0 0 0 rgba(0,167,184,0.26);
  }

  to {
    box-shadow: 0 0 0 3px rgba(0,167,184,0.26);
  }
}

.emotion-9 {
  position: relative;
  top: 50%;
  background-position: center;
  background-size: 100% 2px;
  background-repeat: no-repeat;
  height: 2px;
  border-radius: 2px;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.emotion-9::before {
  content: '';
  position: absolute;
  height: 6px;
  width: 100%;
  -webkit-transform: translateY(-25%);
  -ms-transform: translateY(-25%);
  transform: translateY(-25%);
}

.emotion-2 {
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  right: 0;
  height: 100%;
  width: 100%;
  color: #6C7983;
}

.emotion-0 {
  color: #00c1bb;
}

.emotion-6 {
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  right: 0;
  height: 100%;
  width: 100%;
  color: #00c1bb;
}

.emotion-4 {
  color: #6C7983;
}

.emotion-13 {
  color: #00c1bb;
  top: 50%;
  left: 0;
  position: absolute;
  width: 6px;
  height: 6px;
  margin-top: -3px;
  cursor: default;
  z-index: 2;
}

.emotion-13::before {
  content: '';
  display: block;
  background-color: currentColor;
  width: 6px;
  height: 6px;
  -webkit-transform: scale(2);
  -ms-transform: scale(2);
  transform: scale(2);
  border-radius: 50%;
  -webkit-transition: -webkit-transform 0.2s ease-in-out,width 0.2s ease-in-out,height 0.2s ease-in-out;
  -webkit-transition: transform 0.2s ease-in-out,width 0.2s ease-in-out,height 0.2s ease-in-out;
  transition: transform 0.2s ease-in-out,width 0.2s ease-in-out,height 0.2s ease-in-out;
}

.emotion-13:active::before {
  -webkit-transform: scale(3);
  -ms-transform: scale(3);
  transform: scale(3);
  -webkit-transition-delay: 0s;
  transition-delay: 0s;
}

.emotion-13:focus {
  outline: 0;
}

.emotion-13:focus-visible::before {
  -webkit-animation: animation-0 1s infinite alternate ease-in-out;
  animation: animation-0 1s infinite alternate ease-in-out;
}

<InputRange
  name="x"
>
  <x-range
    disabled={false}
    max={100}
    min={0}
    onChange={[Function]}
    step={1}
  >
    <Track
      className="emotion-11 emotion-8"
      disabled={false}
      style={
        Object {
          "zIndex": 0,
        }
      }
    >
      <div
        className="emotion-8 emotion-9 emotion-10"
        disabled={false}
        style={
          Object {
            "zIndex": 0,
          }
        }
      >
        <Segment
          as={[Function]}
          key="2"
          on={true}
        >
          <svg
            className="emotion-2 emotion-3"
            height="100"
            width="100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <Rect
              disabled={false}
              offset={0}
              width={0}
            >
              <rect
                className="emotion-0 emotion-1"
                disabled={false}
                fill="currentColor"
                height={2}
                offset={0}
                rx={1}
                ry={1}
                width={0}
                x={0}
              />
            </Rect>
          </svg>
        </Segment>
        <Segment
          as={[Function]}
          key="3"
        >
          <svg
            className="emotion-6 emotion-3"
            height="100"
            width="100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <Rect
              disabled={true}
              offset={0}
              width={1000}
            >
              <rect
                className="emotion-4 emotion-1"
                disabled={true}
                fill="currentColor"
                height={2}
                offset={0}
                rx={1}
                ry={1}
                width={1000}
                x={0}
              />
            </Rect>
          </svg>
        </Segment>
      </div>
    </Track>
    <div
      style={
        Object {
          "position": "absolute",
          "transform": undefined,
          "zIndex": 1,
        }
      }
    >
      <Handle
        disabled={false}
        index={0}
        name="xundefined"
        readOnly={false}
        style={Object {}}
      >
        <div
          className="emotion-13 emotion-14"
          data-action="drag-handle"
          index={0}
          style={Object {}}
          tabIndex={0}
        >
          <input
            disabled={false}
            name="xundefined"
            readOnly={false}
            type="hidden"
          />
        </div>
      </Handle>
    </div>
  </x-range>
</InputRange>
`;
